draft/0002-subdocapi.md

```diff
@@ -0,0 +1,380 @@
+exist at the top level as well.
+
+Error codes are common to most operations.
+
+### `get_in(docid, path)`
+Gets a single path from the document. Returns a `Document`
+(or `DocumentFragment`) containing the path value.
+
+* `docid`: Document ID
+* `path`: The sub-document path
+
+### `exists_in(docid, path)`
+
+Like `get_in`, but does not return the value (only checks if it exists)`. If the
+item does not exist, the error code is suitably set in the result (or whatever
+the appropriate error delivery mechanism is for the SDK).
+
+### `upsert_in(docid, path, value, create_parents, cas, ttl, persist_to, replicate_to)`
+Unconditionally places the specified value in the provided path.
+
+* `docid`: Document ID
+* `path`: the sub-document path
+* `value`: the sub-document value. This should be anything JSON-serializable.
+  JSON primitives such as true, false, null, strings and numbers are accepted
+  too.
+* `create_parents`: Boolean parameter indicating whether the full parent tree
+  should be created. By default, only the _last_ path element is created if
+  it does not exist.
+* `cas`: Exactly the same meaning as it has in top-level CRUD operations.
+* `ttl`: Expiration time. This has the same meaning as it does in the CRUD
+  operations. The TTL can be set for _all_ mutation methods.
+* `persist_to`, `replicate_to`: Durability requirements. Functions exactly as in CRUD
+
+> *NOTE*: This (And the other sub-document APIs) will never create or remove documents
+> on the server. This means that if the document does not exist, a `KEY_ENOENT`
+> is returned from the server. The document is never created, even if `create_parents`
+> is true
+
+### `insert_in(docid, path, value, create_parents, cas, ttl, persist_to, replicate_to)`
+Exactly like `upsert_in`, except the path must not already exist.
+If it does, `PathExists` is thrown.
+
+### `replace_in(docid, path, value, cas, ttl, persist_to, replicate_to)`
+Exactly like `upsert_in`, except the penultimate path must exist. If it 
+not exist, `PathNotFound` is thrown. Unlike `insert_in` and `upsert_in`,
+`replace_in` can work on both dictionary and array values
+
+### `extend_in(docid, path, value_or_values, direction, create_parents, cas, ttl, persist_to, replicate_to)`
+Add an element to the back or front of an array.
+
+* `value_or_values`: Can be a single JSON value, or a list of multiple
+  JSON values. The list should be of a special type (MultiValue) so the
+  SDK should not serialize the list itself as JSON.
+* `direction`: `FRONT` or `BACK` depending on whether the item is to be
+  added to the beginning or end of the array
+* `create_parents`: Create the enclosing array and any intermediate elements.
+  If this is not set, then the immediate array must exist.
+
+### `arrayinsert_in(docid, path, value_or_values, cas, ttl, persist_to, replicate_to)`
+Insert an element(s) at a given position in the array. The path should indicate
+the desired position at which to insert the new element(s).
```

Comment:
No, because there is only one case in which create_parents makes sense, and this is when the path is `[0]`, which is equivalent to a prepend anyway.
> On Dec 30, 2015, at 1:53 AM, Michael Nitschinger <notifications@github.com> wrote:
> > +### `arrayinsert_in(docid, path, value_or_values, cas, ttl, persist_to, replicate_to)`
> > +Insert an element(s) at a given position in the array. The path should indicate
> > +the desired position at which to insert the new element(s).
> no create_parents here?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/couchbaselabs/sdk-rfcs/pull/2/files#r48595615>.
>